src/pytorch_lightning/utilities/data.py

```diff
@@ -322,12 +323,54 @@ def _dataloader_init_kwargs_resolve_sampler(
     batch_sampler = getattr(dataloader, "batch_sampler")
     is_predicting = mode == RunningStage.PREDICTING
     # checking the batch sampler type is different than PyTorch default.
-    if batch_sampler is not None and (type(batch_sampler) is not BatchSampler or is_predicting):
-        batch_sampler = type(batch_sampler)(
-            sampler,
-            batch_size=batch_sampler.batch_size,
-            drop_last=(False if is_predicting else batch_sampler.drop_last),
-        )
+    batch_sampler_cls = type(batch_sampler)
+    if batch_sampler is not None and (batch_sampler_cls is not BatchSampler or is_predicting):
+        if hasattr(batch_sampler, "__pl_saved_args"):
+            args = batch_sampler.__pl_saved_args
+            kwargs = batch_sampler.__pl_saved_kwargs
+            arg_names = batch_sampler.__pl_saved_arg_names
+
+            if is_predicting:
+                success, args, kwargs = _replace_value_in_saved_args("drop_last", False, args, kwargs, arg_names)
+                if not success:
+                    rank_zero_warn(
+                        f"Trying to inject `drop_last=False` into batch sampler since you are predicting, however it "
+                        f"seems the class `{batch_sampler_cls.__qualname__}` does not support it. "
+                        "Your predictions might be incomplete. To mitigate this, expose `drop_last` in the `__init__` "
+                        "method of your custom class."
+                    )
```

Comment:
This is not something we do for the DataLoader I hope, right? Because there the drop_last argument is False by default. 

For the batch sampler, I don't fully understand the reasoning here. [The default value is not set](https://pytorch.org/docs/stable/data.html#torch.utils.data.BatchSampler), which means the user has to always provide the value explicitly. Why are we trying to change it then? 

So if my logic is correct here, I would strongly suggest not to interfere with the value for drop_last here, and only send out a warning.

Reply:
Edit: I just saw that we already had this logic before
```
if batch_sampler is not None and (type(batch_sampler) is not BatchSampler or is_predicting):
        batch_sampler = type(batch_sampler)(
            sampler,
            batch_size=batch_sampler.batch_size,
            drop_last=(False if is_predicting else batch_sampler.drop_last),
        )
```
I disagree with this and will open a separate issue about it. So ignore my comment for now as it does not concern this PR.